bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringConfiguration.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigtable.mirroring.hbase1_x;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.regex.Pattern;
+import org.apache.hadoop.conf.Configuration;
+
+public class MirroringConfiguration extends Configuration {
+  Configuration primaryConfiguration;
+  Configuration secondaryConfiguration;
+  MirroringOptions mirroringOptions;
+
+  /**
+   * Key to set to a name of Connection class that should be used to connect to primary database. It
+   * is used as hbase.client.connection.impl when creating connection to primary database.
+   */
+  public static final String MIRRORING_PRIMARY_CONNECTION_CLASS_KEY =
+      "google.bigtable.mirroring.primary-client.connection.impl";
+
+  /**
+   * Key to set to a name of Connection class that should be used to connect to secondary database.
+   * It is used as hbase.client.connection.impl when creating connection to secondary database.
+   */
+  public static final String MIRRORING_SECONDARY_CONNECTION_CLASS_KEY =
+      "google.bigtable.mirroring.secondary-client.connection.impl";
+
+  /**
+   * By default all parameters from the Configuration object passed to
+   * ConnectionFactory#createConnection are passed to Connection instances. If this key is set, then
+   * only parameters that start with given prefix are passed to primary Connection. Use it if
+   * primary and secondary connections' configurations share a key that should have different value
+   * passed to each of connections, e.g. zookeeper url.
+   *
+   * <p>Prefixes should not contain dot at the end.
+   */
+  public static final String MIRRORING_PRIMARY_CONFIG_PREFIX_KEY =
+      "google.bigtable.mirroring.primary-client.prefix";
+
+  /**
+   * If this key is set, then only parameters that start with given prefix are passed to secondary
+   * Connection.
+   */
+  public static final String MIRRORING_SECONDARY_CONFIG_PREFIX_KEY =
```

Comment:
I am not sure I understand the user experience here. Lets say we have an HBase primary and a Bigtable secondary (most common setup), then all the timeouts etc will be common for both, HBase will require zookeeper config while Bigtable will require projectId/InstanceId.  In such scenario you will set all the configs without any prefix (primary or secondary) set.  Each connection impl will take what it wants from the config and everything will work. 

However, if say, we had to setup different timeout values, then we need to specify a prefix (either on primary or secondary), Does that mean that we have to copy all the configurations on that database? A full copy of un-prefixed configs and then a prefixed version with all the configs (including some that will differ)?

Reply:
Yes, as long as all the keys that are common for both clients are the same then no prefix is needed. If The user would want to specify separate configuration for both the databases, we want them to make a copy and keep a prefix. We wanted the user to be explicit about which database is configured to prevent accidental misconfiguration.